### PR TITLE
components: Promote ui/Flex and deprecate `isReversed` prop

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Description
Promotes `ui/Flex` into the canonical `Flex` component for WordPress/components. Also deprecates the `isReversed` prop.

This conflicts with #30261 but replaces the need for it as the new component supports the direction prop.

## How has this been tested?
Run Gutenberg and make sure things still work, especially block dragging and other things that use Flex (angle picker control for example).

## Types of changes
Non-breaking changes.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [N/A] My code follows the accessibility standards. 
- [N/A] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/components/src/flex/README.md [hunk 1]
```diff
@@ -1,98 +1,65 @@
...
+`Flex` is a primitive layout component that adaptively aligns child content horizontally or vertically. `Flex` powers components like `HStack` and `VStack`.
 
 ## Usage
 
+`Flex` is used with any of it's two sub-components, `FlexItem` and `FlexBlock`.
+
 ```jsx
-import { Flex, FlexItem, FlexBlock } from '@wordpress/components';
-import { Icon, wordpress } from '@wordpress/icons';
+import { Flex, FlexItem, FlexBlock, Text } from '@wordpress/components/ui';
```
Comment (gziolo) [inline packages/components/src/flex/README.md]:
```suggestion
import { Flex, FlexItem, FlexBlock, __experimentalText as Text } from '@wordpress/components';
```

File: packages/components/src/flex/README.md [hunk 2]
```diff
@@ -1,98 +1,65 @@
...
-				<h2>WordPress</h2>
+				<Text>Poetry</Text>
 			</FlexBlock>
 		</Flex>
 	);
-};
+}
 ```
 
-The most common use-case for `<Flex />` would be to automatically align two (or more) child items of varying sizes. By default, `<Flex />` would vertically center align them, and evenly spread the inner content horizontally, starting at the edges.
```
Comment (gziolo) [inline packages/components/src/flex/README.md:25]: Should we keep some parts of the old documentation?

File: packages/components/src/flex/README.md [hunk 3]
```diff
@@ -1,98 +1,65 @@
...
-![Flex item with block](https://make.wordpress.org/design/files/2020/06/flex-item-block.png)
-
-`<Flex />` automatically provides "gap" spacing in-between all child items. This value can be adjusted using the `gap` component prop.
-
-`<FlexItem />` and `<FlexBlock />` can be used in combination.
-
-![Multiple Flex items with Flex block](https://make.wordpress.org/design/files/2020/06/flex-item-block-item.png)
+## Props
 
-## Sub-Components
```
Comment (gziolo) [inline packages/components/src/flex/README.md:54]: Do we get sub-components undocumented with the changes proposed?

In practice, when every component exposed in the public API is in its own folder, we can create a README.md file that gets automatically exposed to the developer portal:

https://developer.wordpress.org/block-editor/reference-guides/components/

Every README file gets a link in the sidebar menu.

Screen Shot 2021-04-29 at 08 23 20

File: packages/components/CHANGELOG.md [hunk 4]
```diff
@@ -2,6 +2,10 @@
 
 ## Unreleased
 
+### New Feature
+
+-   `Flex` and its companion components have been re-written from the ground up and the `isReversed` prop has been deprecated.
```
Comment (gziolo) [inline packages/components/CHANGELOG.md]: How about the following notes in the changelog?

```suggestion
### Deprecation

-   `isReversed` prop in `Flex` component has been deprecated. Use `direction` instead ([#31297](https://github.com/WordPress/gutenberg/pull/31297)).

### Internal

-   `Flex`, `FlexBlock`, and `FlexItem` components have been re-written from the ground up ([#31297](https://github.com/WordPress/gutenberg/pull/31297)).
```

**Review by gziolo — COMMENTED** [review]
This looks good, I had some minor notes related to how we organize code and documentation.

File: packages/components/src/flex/README.md (hunk 3 above)
Comment (sarayourfriend) [reply packages/components/src/flex/README.md:54]: Sure! I can move them into their own folders 🙂

File: packages/components/src/flex/flex-block/README.md [hunk 5]
```diff
@@ -0,0 +1,15 @@
+# FlexBlock
+
+A layout component to contain items of a fixed width within `Flex`.
+
+## Usage
+
+See [`flex/README.md#usage`](../flex/README.md#usage) for how to use `FlexBlock`.
```
Comment (gziolo) [inline packages/components/src/flex/flex-block/README.md]: We most likely need the absolute path so it also works at https://developer.wordpress.org/block-editor/reference-guides/components/:
```suggestion
See [`flex/README.md#usage`](/packages/componnents/src/flex/README.md#usage) for how to use `FlexBlock`.
```

Although I'm not sure if it's going to get properly replaced.

File: packages/components/src/flex/flex-block/README.md [hunk 6]
```diff
@@ -0,0 +1,15 @@
+# FlexBlock
+
+A layout component to contain items of a fixed width within `Flex`.
+
+## Usage
+
+See [`flex/README.md#usage`](/packages/comonents/src/flex/README.md#usage) for how to use `FlexBlock`.
```
Comment (gziolo) [inline packages/components/src/flex/flex-block/README.md]:
```suggestion
See [`flex/README.md#usage`](/packages/components/src/flex/README.md#usage) for how to use `FlexBlock`.
```

File: packages/components/src/flex/flex-item/README.md [hunk 7]
```diff
@@ -0,0 +1,21 @@
+# FlexItem
+
+A layout component to contain items of a fixed width within `Flex`.
+
+## Usage
+
+See [`flex/README.md#usage`](/packages/comonents/src/flex/README.md#usage) for how to use `FlexItem`.
```
Comment (gziolo) [inline packages/components/src/flex/flex-item/README.md]:
```suggestion
See [`flex/README.md#usage`](/packages/components/src/flex/README.md#usage) for how to use `FlexItem`.
```

**Review by gziolo — APPROVED** [review]

File: packages/components/src/flex/flex-item/README.md [hunk 8]
```diff
@@ -0,0 +1,21 @@
+# FlexItem
```
Comment (gziolo) [inline packages/components/src/flex/flex-item/README.md]:
```suggestion
# FlexItem (Experimental)
```

File: packages/components/src/flex/flex-block/README.md [hunk 9]
```diff
@@ -0,0 +1,15 @@
+# FlexBlock
```
Comment (gziolo) [inline packages/components/src/flex/flex-block/README.md]:
```suggestion
# FlexBlock (Experimental)
```

**Comment by gziolo** [comment]
As far as I can tell, the example in Storybook needs some adjustments:

Screen Shot 2021-04-30 at 11 44 17

`css` prop doesn't work anymore?

**Pull request merged**